enzyme/Enzyme/GradientUtils.cpp

```diff
@@ -2711,43 +2711,71 @@ Constant *GradientUtils::GetOrCreateShadowFunction(EnzymeLogic &Logic,
        cast<IntegerType>(fn->getReturnType())->getBitWidth() < 16))
     retType = DIFFE_TYPE::CONSTANT;
 
-  // TODO re atomic add consider forcing it to be atomic always as fallback if
-  // used in a parallel context
-  auto &augdata = Logic.CreateAugmentedPrimal(
-      fn, retType, /*constant_args*/ types, TLI, TA,
-      /*returnUsed*/ !fn->getReturnType()->isEmptyTy() &&
-          !fn->getReturnType()->isVoidTy(),
-      type_args, uncacheable_args, /*forceAnonymousTape*/ true, AtomicAdd,
-      PostOpt);
-  Constant *newf = Logic.CreatePrimalAndGradient(
-      (ReverseCacheKey){.todiff = fn,
-                        .retType = retType,
-                        .constant_args = types,
-                        .uncacheable_args = uncacheable_args,
-                        .returnUsed = false,
-                        .shadowReturnUsed = false,
-                        .mode = DerivativeMode::ReverseModeGradient,
-                        .freeMemory = true,
-                        .AtomicAdd = AtomicAdd,
-                        .additionalType = Type::getInt8PtrTy(fn->getContext()),
-                        .typeInfo = type_args},
-      TLI, TA,
-      /*map*/ &augdata);
-  if (!newf)
-    newf = UndefValue::get(fn->getType());
-  auto cdata = ConstantStruct::get(
-      StructType::get(newf->getContext(),
-                      {augdata.fn->getType(), newf->getType()}),
-      {augdata.fn, newf});
-  std::string globalname = ("_enzyme_" + fn->getName() + "'").str();
-  auto GV = fn->getParent()->getNamedValue(globalname);
-
-  if (GV == nullptr) {
-    GV = new GlobalVariable(*fn->getParent(), cdata->getType(), true,
-                            GlobalValue::LinkageTypes::InternalLinkage, cdata,
-                            globalname);
+  switch (mode) {
+  case DerivativeMode::ForwardMode: {
+    Constant *newf =
+        Logic.CreateForwardDiff(fn, retType, types, TLI, TA, false, false, mode,
+                                nullptr, type_args, uncacheable_args);
+
+    if (!newf)
+      newf = UndefValue::get(fn->getType());
+
+    std::string globalname = ("_enzyme_" + fn->getName() + "'").str();
```

Comment:
You probably want a different name from reverse to avoid inadvertant name collision.

Reply:
done. also renamed reverse